cub/device/device_memcpy.cuh

```diff
@@ -0,0 +1,159 @@
+   * auto str_ptrs_in = thrust::make_transform_iterator(gather_iterator,
+   *                                                    GetPtrToStringItem{d_string_data_in, d_string_offsets_old});
+   *
+   * // Returns the string size of the i-th string
+   * auto str_sizes = thrust::make_transform_iterator(gather_iterator, GetStringItemSize{d_string_offsets_old});
+   *
+   * // Returns pointers to the output buffer for each string
+   * auto str_ptrs_out = thrust::make_transform_iterator(thrust::make_counting_iterator(0),
+   *                                                     GetPtrToStringItem{d_string_data_out, d_string_offsets_new});
+   *
+   * // Determine temporary device storage requirements
+   * void *d_temp_storage      = nullptr;
+   * size_t temp_storage_bytes = 0;
+   * cub::DeviceMemcpy::Batch(d_temp_storage, temp_storage_bytes, str_ptrs_in, str_ptrs_out, str_sizes, num_strings);
+   *
+   * // Allocate temporary storage
+   * cudaMalloc(&d_temp_storage, temp_storage_bytes);
+   *
+   * // Run batched copy algorithm (used to permute strings)
+   * cub::DeviceMemcpy::Batch(d_temp_storage, temp_storage_bytes, str_ptrs_in, str_ptrs_out, str_sizes, num_strings);
+   *
+   * // d_string_data_out       <-- "ApplesBananasGrapesOrangesTomatoe"
+   * @endcode
+   * @tparam InputBufferIt <b>[inferred]</b> Device-accessible random-access input iterator type providing the pointers
+   * to the source memory buffers
+   * @tparam OutputBufferIt <b>[inferred]</b> Device-accessible random-access input iterator type providing the pointers
+   * to the destination memory buffers
+   * @tparam BufferSizeIteratorT <b>[inferred]</b> Device-accessible random-access input iterator type providing the
+   * number of bytes to be copied for each pair of buffers
+   * @param d_temp_storage [in] Device-accessible allocation of temporary storage.  When NULL, the required allocation
+   * size is written to \p temp_storage_bytes and no work is done.
+   * @param temp_storage_bytes [in,out] Reference to size in bytes of \p d_temp_storage allocation
+   * @param input_buffer_it [in] Device-accessible iterator providing the pointers to the source memory buffers
+   * @param output_buffer_it [in] Device-accessible iterator providing the pointers to the destination memory buffers
+   * @param buffer_sizes [in] Device-accessible iterator providing the number of bytes to be copied for each pair of
+   * buffers
+   * @param num_buffers [in] The total number of buffer pairs
+   * @param stream [in] <b>[optional]</b> CUDA stream to launch kernels within. Default is stream<sub>0</sub>.
+   * @param debug_synchronous [in] <b>[optional]</b> Whether or not to synchronize the stream after every kernel launch
+   * to check for errors. May cause significant slowdown.  Default is \p false.
+   */
+  template <typename InputBufferIt, typename OutputBufferIt, typename BufferSizeIteratorT>
+  CUB_RUNTIME_FUNCTION static cudaError_t Batch(void *d_temp_storage,
```

Comment:
tbh, I don't think `cub::DeviceMempcy::Batch` nor `cub::DeviceMemcpy::Batched` are the right names here.

I've been wondering for a while how we ended up with the `Batch` terminology when the rest of the CUB algorithms use `Segment`.

Seems like this should be `cub::DeviceSegementedMemcpy`?

Furthermore, the pattern of CUB using a `struct` for the top-level algorithm entity with nested `static` functions is awkward here because it requires us to have a nested name instead of just being able to call `cub::DeviceSegmentedMemcpy(...)` directly.

If we need to keep the type/static function pattern for the sake of consistency, then I would suggest `cub::DeviceSegmentedMempcy::Copy`.

I think the inner name should be a verb. `Batch` and `Batched` are not verbs.

Reply:
Looking back at @allisonvacanti 's [comment on the original issue](https://github.com/NVIDIA/cub/issues/297#issuecomment-837469457) she made the point that `Segment` in CUB usually means "offsets into the same buffer".

The question to ask our selves is "If we were to add an equivalent of DeviceSegmentedSort that worked on independent pointers/sizes, what would we call it?"

My suspicion is we would just add that as a new overload of `DeviceSegmentedSort::SortKeys` instead of adding a new `DeviceBatchSort::SortKeys`.

Reply:
> "If we were to add an equivalent of DeviceSegmentedSort that worked on independent pointers/sizes, what would we call it?"

@jrhemstad I would call it `DeviceBatchedSort`. My understanding is following: 
- segment: a part of something that is separate from the other parts
- batch: a number of things that are dealt with as a group

Segment for me emphases that there's a single array that has to be treated as a number of non-overlapping segments. Batched seems to mean a number of independent parts. These kinds of algorithms have district API - pointer + offsets vs number of pointers. We could call it segmented and pretend that buffers are segments of global address space, but the API is still different and I think that mixing this is confusing. Potential solution might probably be not to specify this at all, just:

```cpp
cub::DeviceMemcpy::Copy(src, dst); // single problem
cub::DeviceMemcpy::Copy(src, dst, offset_begin, offset_end); // segmented problem
cub::DeviceMemcpy::Copy(srcs, dsts, sizes); // batched problem
```

Reply:
`cub::DeviceMemcpy::Copy` sounds find to me.

Reply:
As per offline discussion, we'll go with: `cub::DeviceMemcpy::Batched`